contracts/stacks/contracts/xcall/xcall-impl.clar

```diff
@@ -0,0 +1,708 @@
+    (type (contract-call? .rlp-decode rlp-decode-uint decoded u0))
+    (data (unwrap-panic (element-at decoded u1)))
+  )
+    (ok {
+      type: type,
+      data: data
+    })
+  )
+)
+
+(define-private (parse-protocol (protocol (buff 2048)))
+  (unwrap-panic (as-max-len? (contract-call? .rlp-decode decode-string protocol) u128))
+)
+
+(define-private (parse-cs-message-request (data (buff 2048)))
+  (let (
+    (decoded (contract-call? .rlp-decode rlp-to-list data))
+    (from (unwrap-panic (as-max-len? (contract-call? .rlp-decode rlp-decode-string decoded u0) u128)))
+    (to (unwrap-panic (as-max-len? (contract-call? .rlp-decode rlp-decode-string decoded u1) u128)))
+    (sn (contract-call? .rlp-decode rlp-decode-uint decoded u2))
+    (type (contract-call? .rlp-decode rlp-decode-uint decoded u3))
+    (msg-data (contract-call? .rlp-decode rlp-decode-buff decoded u4))
+    (protocols-list (contract-call? .rlp-decode rlp-decode-list decoded u5))
+    (protocols (map parse-protocol protocols-list))
+  )
+    (ok {
+      from: from,
+      to: to,
+      sn: sn,
+      type: type,
+      data: msg-data,
+      protocols: protocols
+    })
+  )
+)
+
+(define-private (parse-cs-message-result (data (buff 2048)))
+  (let (
+    (decoded (contract-call? .rlp-decode rlp-to-list data))
+  )
+    (ok {
+      sn: (contract-call? .rlp-decode rlp-decode-uint decoded u0),
+      code: (contract-call? .rlp-decode rlp-decode-uint decoded u1),
+      msg: (if (> (len decoded) u2)
+             (some (contract-call? .rlp-decode rlp-decode-buff decoded u2))
+             none
+           )
+    })
+  )
+)
+
+(define-read-only (verify-success (sn uint))
+  (match (map-get? successful-responses { sn: sn })
+    success-response (ok (get value success-response))
+    (ok false)
+  )
+)
+
+(define-public (execute-call (req-id uint) (data (buff 2048)) (receiver <xcall-receiver-trait>) (common <xcall-common-trait>))
+  (let 
+    (
+      (req (unwrap! (map-get? incoming-messages { req-id: req-id }) ERR_MESSAGE_NOT_FOUND))
+      (from (get from req))
+      (to (get to req))
+      (sn (get sn req))
+      (msg-type (get type req))
+      (stored-data-hash (get data-hash req))
+      (protocols (get protocols req))
+      (parsed-to (unwrap! (parse-network-address to) ERR_INVALID_NETWORK_ADDRESS))
+      (to-account (unwrap! (as-max-len? (get account parsed-to) u128) ERR_INVALID_ACCOUNT))
+      (to-principal (unwrap! (contract-call? .util address-string-to-principal to-account) ERR_ADDRESS_TO_PRINCIPAL_FAILED))
+      (receiver-principal (contract-of receiver))
+    )
+      (asserts! (is-eq (keccak256 data) stored-data-hash) ERR_MESSAGE_NOT_FOUND)
+      (asserts! (is-eq to-principal receiver-principal) ERR_INVALID_RECEIVER)
+      (try! (contract-call? receiver handle-call-message from data protocols common))
+      (emit-call-executed-event req-id CS_MESSAGE_RESULT_SUCCESS "")
+      (map-delete incoming-messages { req-id: req-id })
+      (ok true)
```

Comment:
Missing logic to rollback in case of failure

Reply:
what is it supposed to do if the contract call fails?

Reply:
Respond with a failure message so that it can rollback